js/src/dropdown.js

```diff
@@ -402,11 +400,19 @@ class Dropdown extends BaseComponent {
         continue
       }
 
-      if (event && ((event.type === 'click' &&
-          /input|textarea/i.test(event.target.tagName)) ||
-          (event.type === 'keyup' && event.key === TAB_KEY)) &&
-          dropdownMenu.contains(event.target)) {
-        continue
+      if (event) {
+        const dropdownForm = dropdownMenu.querySelector('form')
+
+        if ([context._element, dropdownForm].some(element => event.composedPath().includes(element))) {
+          continue
+        }
+
+        if (((event.type === 'click' &&
+            /input|textarea/i.test(event.target.tagName)) ||
+            (event.type === 'keyup' && event.key === TAB_KEY)) &&
+            dropdownMenu.contains(event.target)) {
+          continue
+        }
```

Comment:
I believe this conditional can be simplified further to the following
```javascript
if (event.type === 'keyup' && event.key === TAB_KEY && dropdownMenu.contains(event.target)) {
  continue
}
```
meaning "tab navigation through the dropdown menu shouldn't close the menu". The click event on input and text area checking looks to be left over complexity. _Unless_ you want to allow interacting with inputs _outside_ of forms, in which case it should be kept in and a test case added.

Reply:
So, handling of form is going to be removed in #33389. BTW removing the checks for `input` and `textarea` will definitely require these elements to be inside the form.

Reply:
Maybe the check could be ```  dropdownForm = dropdownMenu.querySelector('form, input, select, textarea')```

Reply:
@caseyjhol 

these changes passes all the tests too, and simplify it a bit.
Either way, I would like to keep some  comments, in order to be able to know in future what are all these checks ;)

![image](https://user-images.githubusercontent.com/22406063/111699427-c5a2d700-8840-11eb-8c08-68d7f616c883.png)